File: goldenflow/domains/healthcare.py
from __future__ import annotations
import re
import polars as pl
from goldenflow.config.schema import GoldenFlowConfig, TransformSpec
from goldenflow.domains.base import DomainPack
from goldenflow.transforms import register_transform

@register_transform(name="npi_validate", input_types=["string"], auto_apply=False, priority=50, mode="series")
def npi_validate(series: pl.Series) -> pl.Series:
    """Validate NPI numbers (10-digit, Luhn check)."""
    def _validate(val):
        if val is None: return None
        digits = re.sub(r"\D", "", str(val))
        if len(digits) != 10: return False
        # Luhn check with prefix 80840
        full = "80840" + digits
        total = 0
        for i, d in enumerate(reversed(full)):
            n = int(d)
            if i % 2 == 1:
                n *= 2
                if n > 9: n -= 9
            total += n
        return total % 10 == 0
    return series.map_elements(_validate, return_dtype=pl.Boolean)

@register_transform(name="icd10_format", input_types=["string"], auto_apply=False, priority=50, mode="series")
def icd10_format(series: pl.Series) -> pl.Series:
    """Standardize ICD-10 codes (uppercase, insert dot after 3rd char)."""
    def _fmt(val):
        if val is None: return None
        code = val.strip().upper().replace(".", "")
        if len(code) > 3:
            return code[:3] + "." + code[3:]
        return code
    return series.map_elements(_fmt, return_dtype=pl.Utf8)

PACK = DomainPack(
    name="healthcare",
    description="MRN normalization, ICD-10 formatting, NPI validation, date standardization",
    transforms=["npi_validate", "icd10_format", "date_iso8601", "null_standardize", "strip"],
    default_config=GoldenFlowConfig(
        transforms=[
            TransformSpec(column="npi", ops=["npi_validate"]),
            TransformSpec(column="icd10_code", ops=["icd10_format"]),
            TransformSpec(column="service_date", ops=["date_iso8601"]),
            TransformSpec(column="patient_name", ops=["strip", "title_case"]),
        ]
    ),
)

File: packages/goldenflow-js/src/core/domains/healthcare.ts
import type { ColumnValue, DomainPack } from "../types.js";
import { makeConfig } from "../types.js";
import { registerTransform } from "../transforms/registry.js";

function npiValidate(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null || typeof v !== "string") return v;
    const digits = v.replace(/\D/g, "");
    if (digits.length !== 10) return false;
    const full = "80840" + digits;
    let total = 0;
    for (let i = full.length - 1, pos = 0; i >= 0; i--, pos++) {
      let n = parseInt(full[i]!, 10);
      if (pos % 2 === 1) {
        n *= 2;
        if (n > 9) n -= 9;
      }
      total += n;
    }
    return total % 10 === 0;
  });
}

function icd10Format(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null || typeof v !== "string") return v;
    const code = v.trim().toUpperCase().replace(/\./g, "");
    return code.length > 3 ? code.slice(0, 3) + "." + code.slice(3) : code;
  });
}

registerTransform(
  { name: "npi_validate", inputTypes: ["string"], priority: 50, mode: "series" },
  npiValidate,
);
registerTransform(
  { name: "icd10_format", inputTypes: ["string"], priority: 50, mode: "series" },
  icd10Format,
);

export const PACK: DomainPack = {
  name: "healthcare",
  description: "MRN normalization, ICD-10 formatting, NPI validation, date standardization",
  transforms: ["npi_validate", "icd10_format", "date_iso8601", "null_standardize", "strip"],
  defaultConfig: makeConfig({
    transforms: [
      { column: "npi", ops: ["npi_validate"] },
      { column: "icd10_code", ops: ["icd10_format"] },
      { column: "service_date", ops: ["date_iso8601"] },
      { column: "patient_name", ops: ["strip", "title_case"] },
    ],
  }),
};
